### ISSUE TITLE
Typed tuple unpacking in function with overload

### ISSUE THREAD

**Issue by JonathanPlasse** [microsoft/pyright#4288]
**Describe the bug**
An error is raised for each function argument when using a  union of tuples to type a tuple to unpack inside a function with multiple signatures.

**To Reproduce**

I used pyright basic mode.

```python
from socket import socket
from typing import Iterable


def configure_socket(
    socket_options: Iterable[tuple[int, int, int | bytes] | tuple[int, int, None, int]]
) -> None:
    for socket_option in socket_options:
        socket.setsockopt(*socket_option)
```
gives the errors
```txt
error: Argument of type "int | bytes | None" cannot be assigned to parameter "__level" of type "int" in function "setsockopt"
  Type "int | bytes | None" cannot be assigned to type "int"
    "bytes" is incompatible with "int" (reportGeneralTypeIssues)
error: Argument of type "int | bytes | None" cannot be assigned to parameter "__optname" of type "int" in function "setsockopt"
  Type "int | bytes | None" cannot be assigned to type "int"
    "bytes" is incompatible with "int" (reportGeneralTypeIssues)
error: Argument of type "int | bytes | None" cannot be assigned to parameter "__value" of type "None" in function "setsockopt"
  Type "int | bytes | None" cannot be assigned to type "None"
    Type cannot be assigned to type "None" (reportGeneralTypeIssues)
error: Argument of type "int | bytes | None" cannot be assigned to parameter "__optlen" of type "int" in function "setsockopt"
  Type "int | bytes | None" cannot be assigned to type "int"
    "bytes" is incompatible with "int" (reportGeneralTypeIssues)
```

**Expected behavior**
`socket.setsockopt` has these type signatures:
```python
setsockopt(self: socket, __level: int, __optname: int, __value: int | bytes, /) -> None
setsockopt(self: socket, __level: int, __optname: int, __value: None, __optlen: int, /) -> None
```
So, unpacking `tuple[int, int, int | bytes] | tuple[int, int, None, int]` should not raise errors. as each tuple match one of the signature.

**VS Code extension or command-line**
I use pyright 1.1.282 with npx.

**Comment by erictraut** [comment]
This won't work given the way that pyright implements overload matching. Pyright will "expand" unions for individual arguments, but this is done before unpacking occurs. You'll need to find a workaround if you want this to type check in pyright.
